libwebauthn/src/transport/ble/btleplug/connection.rs

```diff
@@ -0,0 +1,147 @@
+
+use btleplug::api::{Peripheral as _, WriteType};
+use btleplug::platform::Peripheral;
+use byteorder::{BigEndian, ReadBytesExt};
+use tracing::{debug, info, instrument, trace, warn};
+
+use super::device::FidoEndpoints;
+use super::Error;
+use crate::fido::FidoRevision;
+use crate::transport::ble::framing::{
+    BleCommand, BleFrame as Frame, BleFrameParser, BleFrameParserResult,
+};
+
+#[derive(Debug, Clone)]
+pub struct Connection {
+    pub peripheral: Peripheral,
+    pub services: FidoEndpoints,
+}
+
+impl Connection {
+    pub async fn new(
+        peripheral: &Peripheral,
+        services: &FidoEndpoints,
+        revision: &FidoRevision,
+    ) -> Result<Self, Error> {
+        let connection = Self {
+            peripheral: peripheral.to_owned(),
+            services: services.clone(),
+        };
+        connection.select_fido_revision(revision).await?;
+        Ok(connection)
+    }
+
+    async fn control_point_length(&self) -> Result<usize, Error> {
+        let max_fragment_length = self
+            .peripheral
+            .read(&self.services.control_point_length)
+            .await
+            .or(Err(Error::OperationFailed))?;
+
+        if max_fragment_length.len() != 2 {
+            warn!(
+                { len = max_fragment_length.len() },
+                "Control point length endpoint returned an unexpected number of bytes",
+            );
+            return Err(Error::OperationFailed);
+        }
+
+        let mut cursor = IOCursor::new(max_fragment_length);
+        let max_fragment_size = cursor.read_u16::<BigEndian>().unwrap() as usize;
+        Ok(max_fragment_size)
+    }
+
+    pub async fn frame_send(&self, frame: &Frame) -> Result<(), Error> {
+        let max_fragment_size = self.control_point_length().await?;
+        let fragments = frame
+            .fragments(max_fragment_size)
+            .or(Err(Error::InvalidFraming))?;
+
+        for (i, fragment) in fragments.iter().enumerate() {
+            debug!({ fragment = i, len = fragment.len() }, "Sending fragment");
+            trace!(?fragment);
+
+            self.peripheral
+                .write(
+                    &self.services.control_point,
+                    fragment,
+                    WriteType::WithoutResponse,
+                )
+                .await
+                .or(Err(Error::OperationFailed))?;
+        }
+
+        Ok(())
+    }
+
+    pub(crate) async fn select_fido_revision(&self, revision: &FidoRevision) -> Result<(), Error> {
+        let ack: u8 = revision.clone() as u8;
+        self.peripheral
+            .write(
+                &self.services.service_revision_bitfield,
+                &[ack],
+                WriteType::WithoutResponse,
+            )
+            .await
+            .or(Err(Error::OperationFailed))?;
+
+        info!(?revision, "Successfully selected FIDO revision");
+        Ok(())
+    }
+
+    #[instrument(skip_all)]
+    pub async fn frame_recv(&self) -> Result<Frame, Error> {
```

Comment:
In theory, we could add a timeout to this recv-function at least, and watch it ourselves, as btleplug doesn't do custom timeouts. But I'm not sure it's worth it, as it wouldn't really be what we want. If the `update()` hangs internally, us watching outside wouldn't help. Timekeeping would also rely on the Keepalives or other updates to periodically come in.